Rubocop markdown snippets

zzak opened the pull request:
Introduce [rubocop-md](https://github.com/rubocop/rubocop-md) for our markdown files.

This ensures any code snippets in the guides also follow the Rails coding conventions, without having to manually review every change that comes in.

I've intentionally disabled `Style/StringLiterals` in the guides because that is a much larger change, and figured I would follow up with a single PR to do that after this gets merged.

You can see some of the changes broken down further in zzak/rails#2 and zzak/rails#3.

matthewd commented:
I like the idea, though there are some specifics that don't seem great (`attr_reader`, `%Q`) -- being documentation examples, it can be better for us to use more generic constructs than optimise for the actual [dummy] content that's present as shown.

I also spotted a `rubocop:disable` comment somewhere.. does that show up in the generated docs?

zzak commented:
Reverted some of the extraneous rules in 3ffc441.

> I also spotted a rubocop:disable comment somewhere.. does that show up in the generated docs?

It did show up, so I reverted that as well. In spite of the [docs](https://github.com/rubocop/rubocop-md#but-if-i-want-to-use-inline-disabling-some-directive-like-in-classic-rubocop) providing an alternative method (`<span style="display:none;"># rubocop:disable all</span>`) that did not work when specifying this specific rule.. so probably better to just disable it.

We probably shouldn't worry about redundant begin/end in our docs 😂 and fine if we have to catch that one in code review IMO.

zzak commented on guides/source/active_record_validations.md:
````diff
@@ -1073,7 +1073,7 @@ Passing an array of symbols is also acceptable.
 ```ruby
 class Book
-  include ActiveModel:Validations
+  include ActiveModel::Validations
````
Regardless of the style consistency wins, the fact this does a `Ripper.sexp` to check the syntax of the code caught this which is a huge benefit IMO.

sambostock commented on guides/source/action_mailbox_basics.md:
```diff
@@ -272,8 +272,8 @@ Configure basic routing:
 class ApplicationMailbox < ActionMailbox::Base
-  routing /^save@/i     => :forwards
-  routing /@replies\./i => :replies
+  routing (/^save@/i     => :forwards)
+  routing (/@replies\./i => :replies)
```
I _think_ this is a syntax error because of the space between `routing` and `(`.

If I'm correct, I'm surprised it isn't caught by RuboCop, especially given https://github.com/rails/rails/pull/47186#discussion_r1091430261... 🤔

Was this autocorrected?

zzak replied:
@sambostock Thanks for taking a look!

This is not a syntax error, although I would prefer `routing(...)`, it doesn't seem like there is a rule for that at least in our config.

```ruby
>> RUBY_VERSION
=> "3.3.0"
?> def foo bar
?>   puts bar
>> end
=> :foo
>>
>> foo "omg"
omg
=> nil
>> foo("omg")
omg
=> nil
>> foo ("omg")
omg
=> nil
```

koic replied:
I think the syntax error that @sambostock tells is as follows:

## With space

invalid syntax:

```console
% ruby -vce 'routing (/^save@/i     => :forwards)'
ruby 3.2.0 (2022-12-25 revision a528908271) [x86_64-darwin19]
-e:1: void value expression
...ing (/^save@/i     => :forwards)
-e: compile error (SyntaxError)
```

## No space

valid syntax:

```console
% ruby -vce 'routing(/^save@/i     => :forwards)'
ruby 3.2.0 (2022-12-25 revision a528908271) [x86_64-darwin19]
Syntax OK
```

The same is true for Ruby 3.3.0-dev.

zzak replied:
Thanks! I will fix it.

zzak replied:
@sambostock @koic Fixed in 321c8df, but I wonder why `Ripper.sexp` didn't catch this? Is there another rule we can add that would have caught it?

koic replied:
I'm not sure, but maybe there is something wrong with the Parser gem. Unfortunately, the cause has not been determined.

palkan replied:
Both Ripper and Parser recognize this syntax:

```ruby
irb(main):003:0> Ripper.sexp('routing (/^save@/i     => :forwards)')
=> 
[:program,
   [:@ident, "routing", [1, 0]],
   [:args_add_block,
    [[:paren,
      [:case,
       [:regexp_literal, [[:@tstring_content, "^save@", [1, 10]]], [:@regexp_end, "/i", [1, 16]]],
       [:in, [:symbol_literal, [:symbol, [:@ident, "forwards", [1, 27]]]], nil, nil]]]],
    false]]]]

irb(main):006:0> Parser::CurrentRuby.parse('routing (/^save@/i     => :forwards)')
=> 
s(:send, nil, :routing,                                                      
  s(:begin,                                                                  
    s(:match_pattern,                                                
      s(:regexp,                                                     
        s(:str, "^save@"),                                           
        s(:regopt, :i)),                                             
      s(:sym, :forwards)))) 
```

But:

```sh
$ ruby -w -e 'routing (/^save@/i     => :forwards)'

-e:1: void value expression
...ing (/^save@/i     => :forwards)
-e: compile error (SyntaxError)
```

🤷‍♂️

rafaelfranca commented on guides/.rubocop.yml:
```diff
@@ -0,0 +1,23 @@
+  Enabled: false
+
+Style/RedundantPercentQ:
+  Enabled: false
```
Let's add it.

rafaelfranca commented on guides/source/association_basics.md:
````diff
@@ -2857,7 +2853,7 @@ end
 With this definition complete, our `Entry` delegator now provides the following methods:
 
-```ruby
+```
````
Why isn't this ruby anymore?

zzak replied:
It's parseable Ruby in that every `Entry` is a thing, but `Entry#whatever` will just be a comment after that. I think by specifying ruby as the code block, it means this code must adhere to our style guidelines.

I've also considered rewriting this as a table, because I prefer that to code for readability, e.g. (patch later):

![Screenshot 2023-02-02 at 10 46 21](https://user-images.githubusercontent.com/277819/222856794-0e13cc83-fe81-4b55-b334-225700cf2e95.png)

rafaelfranca replied:
Table would be a good solution for this.

zzak replied:
Resolved in #47653

zzak committed:
Add rubocop-md for linting guides snippets
